### DISCUSSION TITLE
TypeAlias that depends on library version in environment

### DISCUSSION THREAD

**Discussion by smackesey** [microsoft/pyright#4236] [Q&A]
I work on a library that needs to support the time manipulation library [Pendulum](https://pendulum.eustace.io/) in both versions 1.x and 2.x.

There was a breaking change between v1 and v2 of pendulum where a key class was moved from `pendulum.Pendulum` to `pendulum.DateTime`. The relevant parts of the public interface of the class remain unchanged.

This is causing difficulty with type annotations-- I'd like to create a type alias `PendulumDateTime` that points to the appropriate class depending on the environment. Something like this:

```python
import packaging.version
import pendulum

if packaging.version.parse(pendulum.__version__).major >= 2:
    PendulumDateTime: TypeAlias = pendulum.DateTime
else:
    PendulumDateTime: TypeAlias = pendulum.Pendulum
```

Of course this doesn't work, since the assignment depends on runtime logic. Is there any way around this?

**Comment by erictraut** [comment]
This may be better asked in the [python/typing discussion forum](https://github.com/python/typing/discussions) since it's not specific to pyright, but I'll offer my suggestion here.

I don't see a great way to handle this in static typing. My inclination would be to create a wrapper class that hides as much of the difference as possible and provides the runtime version checks that select between the external types. If it is necessary to expose the underlying pendulum type from this wrapper class, it would need to be exposed as a union of `DateTime` and `Pendulum`.

Another option is to drop support for pendulum v1 and require v2 with your library. Not sure if that's feasible.
